### PR TITLE
Run nfd-master as Deployment

### LINKED ISSUE #4: Run nfd-master as Deployment
Run nfd-master as Deployment instead of a DaemonSet, similar what the [NFD upstream deployment template](https://github.com/kubernetes-sigs/node-feature-discovery/blob/master/nfd-master.yaml.template) does.

Makes it possible to deploy NFD in managed clusters which have no master nodes.

### PR DESCRIPTION
**Pull request by ArangoGutierrez**
Fixes #4 

Also harden RBAC requested by the controller

### CODE REVIEW

**Comment by marquiz** [comment]
@ArangoGutierrez could the RBAC rule changes be split out into a separate PR? At least those that are not required by the Deployment change.

File: Makefile [hunk 1]
```diff
@@ -70,7 +70,7 @@ BASE_IMAGE_MINIMAL ?= gcr.io/distroless/base
 IMAGE_TAG_RBAC_PROXY ?= gcr.io/kubebuilder/kube-rbac-proxy:v0.8.0
 
 # Produce CRDs that work back to Kubernetes 1.11 (no version conversion)
-CRD_OPTIONS ?= "crd:trivialVersions=true,preserveUnknownFields=false"
+CRD_OPTIONS ?= "crd"
```
Comment (marquiz) [inline Makefile:73]: what's this for?

File: config/crd/bases/nfd.kubernetes.io_nodefeaturediscoveries.yaml [hunk 2]
```diff
@@ -107,13 +106,12 @@ spec:
...
-                    \    // +patchStrategy=merge     // +listType=map     // +listMapKey=type
-                    \    Conditions []metav1.Condition `json:\"conditions,omitempty\"
-                    patchStrategy:\"merge\" patchMergeKey:\"type\" protobuf:\"bytes,1,rep,name=conditions\"`
-                    \n     // other fields }"
+                    type FooStatus struct{ // Represents the observations of a foo's
+                    current state. // Known .status.conditions.type are: \"Available\",
+                    \"Progressing\", and \"Degraded\" // +patchMergeKey=type // +patchStrategy=merge
+                    // +listType=map // +listMapKey=type Conditions []metav1.Condition
+                    `json:\"conditions,omitempty\" patchStrategy:\"merge\" patchMergeKey:\"type\"
+                    protobuf:\"bytes,1,rep,name=conditions\"` \n // other fields }"
```
Comment (marquiz) [inline config/crd/bases/nfd.kubernetes.io_nodefeaturediscoveries.yaml:114]: nit: is there something significant here?

File: config/default/kustomization.yaml [hunk 3]
```diff
@@ -13,14 +13,15 @@ namespace: node-feature-discovery-operator
...
 # If you want your controller-manager to expose the /metrics
 # endpoint w/o any authn/z, please comment the following line.
+patchesStrategicMerge:
 - manager_auth_proxy_patch.yaml
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+resources:
+- ../crd
+- ../rbac
+- ../manager
```
Comment (marquiz) [inline config/default/kustomization.yaml:27]: nit: needed for this?

File: config/rbac/role.yaml [hunk 4]
```diff
@@ -105,14 +104,6 @@ rules:
...
   - update
   - watch
-- apiGroups:
-  - ""
-  resources:
-  - nodes/status
-  verbs:
-  - get
-  - patch
-  - update
```
Comment (marquiz) [inline config/rbac/role.yaml:115]: Why are these removed? Updating extended resources requires this but I'm not sure if we support those.

**Review by marquiz — COMMENTED** [review]
Thanks @ArangoGutierrez. Some quick nits about the deployment files 😄

File: controllers/nodefeaturediscovery_controller.go [hunk 5]
```diff
@@ -103,6 +103,7 @@ func validateUpdateEvent(e *event.UpdateEvent) bool {
 // +kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=core,resources=pods/log,verbs=get
 // +kubebuilder:rbac:groups=apps,resources=daemonsets,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=apps,resources=deployments,verbs=get;list;watch;create;update;patch;delete
```
Comment (marquiz) [inline controllers/nodefeaturediscovery_controller.go:106]: Should we remove the daemonsert line above? Or is it needed for updates or smth?

File: controllers/nodefeaturediscovery_controls.go [hunk 6]
```diff
@@ -388,55 +388,113 @@ func DaemonSet(n NFD) (ResourceStatus, error) {
...
+	return Ready, nil
+}
+
+// Deployment checks the readiness of a Deployment and creates one if it doesn't exist
+func Deployment(n NFD) (ResourceStatus, error) {
+	// state represents the resource's 'control' function index
+	state := n.idx
+
+	// It is assumed that the index has already been verified to be a
+	// DaemonSet object, so let's get the resource's Deployment object
```
Comment (marquiz) [inline controllers/nodefeaturediscovery_controls.go:441]: Daemonset??

File: controllers/nodefeaturediscovery_controls.go [hunk 7]
```diff
@@ -388,55 +388,113 @@ func DaemonSet(n NFD) (ResourceStatus, error) {
...
 
-		// Set the args based on the port that was determined
-		// and the instance that was determined
-		obj.Spec.Template.Spec.Containers[0].Args = args
+	if strings.TrimSpace(n.ins.Spec.LabelWhiteList) != "" {
+		args = append(args, fmt.Sprintf("--label-whitelist=%q", n.ins.Spec.LabelWhiteList))
 	}
 
+	// Set the args based on the port that was determined
+	// and the instance that was determined
```
Comment (marquiz) [inline controllers/nodefeaturediscovery_controls.go:486]: This comment could be dropped, doesn't seem to make much sense anymore(?)

File: controllers/nodefeaturediscovery_resources.go [hunk 8]
```diff
@@ -263,6 +275,22 @@ func (r *NodeFeatureDiscoveryReconciler) deleteDaemonSet(ctx context.Context, na
 	return r.Delete(context.TODO(), ds)
 }
 
+// deleteDaemonSet deletes Operand DaemonSet
```
Comment (marquiz) [inline controllers/nodefeaturediscovery_resources.go:278]:
```suggestion
// deleteDeployment deletes Operand DaemonSet
```

File: controllers/nodefeaturediscovery_status.go [hunk 9]
```diff
@@ -282,8 +279,32 @@ func (r *NodeFeatureDiscoveryReconciler) getTopologyUpdaterDaemonSetConditions(c
 
 // getMasterDaemonSetConditions is a wrapper around "getDaemonSetConditions" for
 // master DaemonSets
-func (r *NodeFeatureDiscoveryReconciler) getMasterDaemonSetConditions(ctx context.Context, instance *nfdv1.NodeFeatureDiscovery) (Status, error) {
-	return r.getDaemonSetConditions(ctx, instance, nfdMasterApp)
+func (r *NodeFeatureDiscoveryReconciler) getMasterDeploymentConditions(ctx context.Context, instance *nfdv1.NodeFeatureDiscovery) (Status, error) {
```
Comment (marquiz) [inline controllers/nodefeaturediscovery_status.go:282]: nit: Adjust the comment above, too

File: controllers/nodefeaturediscovery_status.go [hunk 10]
```diff
@@ -282,8 +279,32 @@ func (r *NodeFeatureDiscoveryReconciler) getTopologyUpdaterDaemonSetConditions(c
...
+	return r.getDeploymentConditions(ctx, instance, nfdMasterApp)
+}
+
+// getDaemonSetConditions gets the current status of a DaemonSet. If an error
```
Comment (marquiz) [inline controllers/nodefeaturediscovery_status.go:286]:
```suggestion
// getDeploymentConditions gets the current status of a Deployment. If an error
```

File: controllers/nodefeaturediscovery_status.go [hunk 11]
```diff
@@ -282,8 +279,32 @@ func (r *NodeFeatureDiscoveryReconciler) getTopologyUpdaterDaemonSetConditions(c
...
+	status := initializeDegradedStatus()
+
+	d, err := r.getDeployment(ctx, instance.ObjectMeta.Namespace, nfdAppName)
+	if err != nil {
+		return status, err
+	}
+
+	dStatus := d.Status.DeepCopy()
+
+	if dStatus.AvailableReplicas == 0 {
```
Comment (marquiz) [inline controllers/nodefeaturediscovery_status.go:299]: Just a question: is one replica enough? I think in this PR we always use only one replica so it probably is. Maybe add a TODO note to implement more intelligent logic (and more conditions?) if we make the number of replicas configurable.

File: controllers/nodefeaturediscovery_controller.go (hunk 5 above)
Comment (ArangoGutierrez) [reply controllers/nodefeaturediscovery_controller.go:106]: worker will still be a daemon ;)

File: controllers/nodefeaturediscovery_status.go (hunk 11 above)
Comment (ArangoGutierrez) [reply controllers/nodefeaturediscovery_status.go:299]: 👍🏽  like the idea will add

File: config/rbac/role.yaml (hunk 4 above)
Comment (marquiz) [reply config/rbac/role.yaml:115]: Yeah, I think we definitely need these as we have support for `ResourceLabels`

File: controllers/nodefeaturediscovery_controller.go (hunk 5 above)
Comment (marquiz) [reply controllers/nodefeaturediscovery_controller.go:106]: Well, true 🤦 Sorry...

File: controllers/nodefeaturediscovery_status.go (hunk 10 above)
Comment (marquiz) [reply controllers/nodefeaturediscovery_status.go:286]: Still unchanged

**Review by marquiz — CHANGES_REQUESTED** [review]
I don't think you should change `config/rbac/role.yaml`

**Review by marquiz — APPROVED** [review]
Thanks @ArangoGutierrez! Looks good, now
/lgtm
